### PR TITLE
Update PL-Lightning to support new features in PL

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
Current PL-Lightning is not compatible with new features in PL and as the result almost all `gates` and `adjoint_jacobian` tests are FAILED. This PR updates the PL-lightning along with all tests to support new features in PL. 

**Description of the Change:**
These files are mainly updated: 
- `lightning_qubit.py`
- `tests/test_gates.py`
- `tests/test_adjoint_jacobian.py`

**Benefits:**
PL-Lightning is compatible with new features in PL and there shouldn't be any failed tests 

**Related GitHub Issues:**
n/a

### CODE REVIEW

File: pennylane_lightning/lightning_qubit.py [hunk 1]
```diff
@@ -220,10 +220,11 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
 
         ops_serialized = adj.create_ops_list(*ops_serialized)
 
+        tape_trainable_params_set = set(tape.trainable_params)
```
Comment (josh146) [inline pennylane_lightning/lightning_qubit.py:223]: @maliasadi just something to keep in mind is that the reason we changed `trainable_params` from being a set is that we were affected by a rare python bug where set order of integers is not guaranteed to be stable!

So something to be careful with if you plan to iterate over a set, and expect the integers to be ordered.

Comment (mlxd) [reply pennylane_lightning/lightning_qubit.py:223]: I think it should be fine --- we don't really care about order as much as the existence of indices. We iterate over them in the order given to CPP, and assume no guarantees on being sorted (I think).

Comment (maliasadi) [reply pennylane_lightning/lightning_qubit.py]: Thank you @josh146 @mlxd! I update `AdjointDiff` method to work with `trainable_params` as a list.

File: tests/test_adjoint_jacobian.py [hunk 2]
```diff
@@ -250,18 +252,29 @@ def test_multiple_rx_gradient(self, tol):
...
-        args = np.linspace(0.2, 0.5, op.num_params)
 
+        # op.num_wires and op.num_params must be initialized a priori
         with qml.tape.JacobianTape() as tape:
             qml.Hadamard(wires=0)
             qml.RX(0.543, wires=0)
             qml.CNOT(wires=[0, 1])
 
-            op(*args, wires=range(op.num_wires))
+            op
```
Comment (mlxd) [inline tests/test_adjoint_jacobian.py:277]: Just curious if the operation is being applied here with this? Or is this a GH formatting issue that I cannot see the rest of the changes?

File: tests/test_adjoint_jacobian.py [hunk 3]
```diff
@@ -462,15 +475,13 @@ def circuit(params):
             qml.Rot(params[1], params[0], 2 * params[0], wires=[0])
             return qml.expval(qml.PauliX(0))
 
-        spy_numeric = mocker.spy(qml.tape.JacobianTape, "numeric_pd")
```
Comment (mlxd) [inline tests/test_adjoint_jacobian.py:465]: Should we be checking this?

File: tests/test_adjoint_jacobian.py [hunk 4]
```diff
@@ -462,15 +475,13 @@ def circuit(params):
...
         spy_analytic = mocker.spy(dev, "adjoint_jacobian")
 
         cost = QNode(circuit, dev, diff_method="finite-diff")
 
         grad_fn = qml.grad(cost)
         grad_F = grad_fn(params)
 
-        spy_numeric.assert_called()
```
Comment (mlxd) [inline tests/test_adjoint_jacobian.py:473]: And here?

**Review by mlxd — COMMENTED** [review]
Nice one Ali. Just a few small questions

File: tests/test_adjoint_jacobian.py (hunk 2 above)
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:277]: Previously, `op` was defined as a list of operations' inits and this test function worked properly as values of `op.num_params` and `op.num_wires` were known for each operation. Now, with the recent updates, these values are not known and one must pass arguments a priori. Hence, each operation is being applied here with each `op` call without any needs on passing args and wires.

File: tests/test_adjoint_jacobian.py (hunk 3 above)
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:465]: It seems this is not required anymore as they are also deprecated in PL test-suite.

Comment (josh146) [reply tests/test_adjoint_jacobian.py:465]: Yep, `JacobianTape` (and all the other tape subclasses!) are being deprecated in favour of the functions in the `qml.gradients` subpackage.

Comment (josh146) [reply tests/test_adjoint_jacobian.py:465]: Note that these checks _could_ be updated here, _but_ I don't think it makes sense to unittest PL core functionality in the plugin test.

**Review by mlxd — APPROVED** [review]
:100: Great job @maliasadi

**Pull request merged**